File: rdcm/src/main/java/com/metsr/hpc/MabManager.java
package com.metsr.hpc;
import java.io.BufferedReader;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;
import java.net.URI;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

public class MabManager{
	private Mab mab;
	private MabBus mabBus;
	private HashMap<Integer, ArrayList<Double>> initialLinkSpeedLength;
	private ConcurrentHashMap<Integer, Double> roadLengthMap; //July,2020
	public String data_dir;

	public MabManager(String data_dir_loc){
		HashMap<String, ArrayList<ArrayList<Integer>>> path_info = new HashMap<String, ArrayList<ArrayList<Integer>>>();
		HashMap<String, ArrayList<Integer>> valid_path = new HashMap<String, ArrayList<Integer>>();
		// July,2020
		HashMap<String, ArrayList<ArrayList<Integer>>> path_info_bus = new HashMap<String, ArrayList<ArrayList<Integer>>>();
		HashMap<String, ArrayList<Integer>> valid_path_bus = new HashMap<String, ArrayList<Integer>>();
		mab = new Mab(path_info, valid_path);
		//July, 2020
		mabBus = new MabBus(path_info_bus,valid_path_bus);
		initialLinkSpeedLength = new HashMap<Integer, ArrayList<Double>>();
		//July, 2020
		roadLengthMap = new ConcurrentHashMap<Integer,Double>();
		data_dir = data_dir_loc;
	}

	// it should be called every time tick.  [t1,t2]
	// input is the OD.  t1.
	// output is the route.  t1. 
	// refresh the energy.  t2.
	// t1, t1.
	
	public int ucbRouting(String od){
		mab.play(od);
		return mab.getAction();
	}
	
	// t2.
	public void refreshLinkUCB(ConcurrentHashMap<Integer, ArrayList<Double>> linkUCBMap){
//		synchronized (linkUCBMap) {
		mab.updateLinkUCB(linkUCBMap);
//		}
		
	}	
	
	public void refreshRouteUCB(ConcurrentHashMap<String, List<List<Integer>>> routeUCBMap){
//		synchronized (routeUCBMap) {
		mab.updateRouteUCB(routeUCBMap);
//		}
	}
	
	//July, 2020
	public int ucbRoutingBus(String od){
		mabBus.playBus(od);
		return mabBus.getAction();
	}
	
	//July, 2020 routeUCB bus
	public void refreshRouteUCBBus(ConcurrentHashMap<String, List<List<Integer>>> routeUCBMapBus){
//		synchronized (routeUCBMapBus) {
		mabBus.updateRouteUCBBus(routeUCBMapBus);
//		}
		
	}
	//July, 2020
	public void refreshLinkUCBBus(ConcurrentHashMap<Integer, ArrayList<Double>> linkUCBMapBus){
//		synchronized (linkUCBMapBus){
		mabBus.updateLinkUCBBus(linkUCBMapBus);
//		}
	}
	
	//July, 2020
	public void refreshLinkUCBShadow(ConcurrentHashMap<Integer, ArrayList<Double>> speedUCBMap, Map<Integer, Double> lengthUCB){
//		synchronized (speedUCBMap){
		mabBus.updateShadowBus(speedUCBMap,lengthUCB);
//		}
	}	
	
	public void initializeLinkEnergy1() {
		try {
			/* CSV file for data attribute */
			String fileName1 = data_dir +"/NYC/background_traffic/background_traffic_NYC.csv";
			BufferedReader br = new BufferedReader(new FileReader(fileName1));
			//br.readLine();          //the first row is the title row
			String line = null; 
			br.readLine();
			while ((line = br.readLine())!=null) {
				//line = br.readLine();
				String[] result1 = line.split(",");
				int roadID = Integer.parseInt(result1[0]);
				double SpeedSum = 0.0;
				for (int i = 1; i < 25; i++) {
					SpeedSum += Double.parseDouble(result1[i]);
				}
				double averageSpeed = SpeedSum/24.0;
				ArrayList<Double> speedLength = new ArrayList<Double>();
				speedLength.add(averageSpeed);
				initialLinkSpeedLength.put(roadID, speedLength);	
			}
			br.close();
		}catch (FileNotFoundException e){
			System.out.println("ContextCreator: No speed csv file found");
			e.printStackTrace();
		}catch (IOException e){
	        e.printStackTrace();
	    }
	}
	
	public void initializeLinkEnergy2(){
		try {
			/* CSV file for data attribute */
			String fileName2 = data_dir+"/NYC/background_traffic/background_traffic_NYC.csv";
			BufferedReader br = new BufferedReader(new FileReader(fileName2));
			br.readLine();          //the first row is the title row
			String line = null; 
			while ((line = br.readLine())!=null) {
				//line = br.readLine();
				String[] result=line.split(",");
				int roadID = Integer.parseInt(result[0]);
				double roadLength = Double.parseDouble(result[result.length-1]);
				ArrayList<Double> speedLength = new ArrayList<Double>();
				
				speedLength.add(initialLinkSpeedLength.get(roadID).get(0));
				speedLength.add(roadLength);
				initialLinkSpeedLength.put(roadID, speedLength);
				roadLengthMap.put(roadID,roadLength);  //July,2020
			}
			br.close();
		}catch (FileNotFoundException e){
			System.out.println("ContextCreator: No road csv file found");
			e.printStackTrace();
		}catch (IOException e){
			e.printStackTrace();
		}
		mab.warm_up(initialLinkSpeedLength);
		
		//July, 2020
		mabBus.warm_up_bus(initialLinkSpeedLength);
		//July, 2020
		
		
	}
	
	// July,2020
	public ConcurrentHashMap<Integer, Double> getRoadLengthMap(){
	    	return roadLengthMap;
	}
}

File: eco_routing/MabManager.py
import os
import sys
import json
from os import path
from eco_routing.Mab import MAB
# from eco_routing.MabBus import MABBus

class MABManager(object):
    def __init__(self, working_dir, args):
        self.mab = {} # MAB model for E-Taxis, every hour we train a new model
        # self.mabBus = {} # MAB model for E-Buses
        self.initialLinkSpeedLength = []
        self.roadLengthMap = {}
        self.path_info = {}
        self.valid_path = {}
        self.path_info_bus = {}
        self.valid_path_bus = {}

        self.working_dir = working_dir
        self.args=args
        for hour in range(int(args.SIMULATION_STOP_TIME * args.SIMULATION_STEP_SIZE//3600)+1):
            self.mab[hour] = MAB(self.path_info, self.valid_path)
            # self.mabBus[hour] = MABBus(self.path_info_bus, self.valid_path_bus)
            self.initialLinkSpeedLength.append({})

    def ucbRouting(self, od, hour):
        self.mab[hour].play(od)
        return self.mab[hour].getAction()

    def refreshLinkUCB(self, new_linkUCBMap):
        hour = 0
        for IDhour in new_linkUCBMap.keys():
            hour = int(IDhour.split(";")[1])
        #    break
        #key_hour='hour_int'
        #if key_hour in new_linkUCBMap.keys():
        #   hour=new_linkUCBMap[key_hour]
        self.mab[hour].updateLinkUCB(new_linkUCBMap)
        return hour

    def refreshRouteUCB(self, new_routeUCBMap):
        for hour in range(int(self.args.SIMULATION_STOP_TIME * self.args.SIMULATION_STEP_SIZE//3600)+1):
            self.mab[hour].updateRouteUCB(new_routeUCBMap)

    # def ucbRoutingBus(self, od, hour):
    #     self.mabBus[hour].play(od)
    #     return self.mabBus[hour].getAction()

    # def refreshLinkUCBBus(self, new_linkUCBMapBus):
    #     hour = 0
    #     # TODO: Change to json
    #     for IDhour in new_linkUCBMapBus.keys(): 
    #         #print('hour is')
    #         #print(IDhour)
    #         hour = int(IDhour.split(";")[1])
    #     #    break
 
    #     #key_hour='hour_int'
    #     #if key_hour in new_linkUCBMapBus.keys():
    #     #    hour=new_linkUCBMapBus[key_hour]
    #     self.mabBus[hour].updateLinkUCB(new_linkUCBMapBus)
    #     return hour

    # def refreshRouteUCBBus(self, new_routeUCBMapBus):
    #     for hour in range(int(self.args.SIMULATION_STOP_TIME * self.args.SIMULATION_STEP_SIZE//3600)+1):
    #         self.mabBus[hour].updateRouteUCB(new_routeUCBMapBus)

    # def refreshLinkUCBShadow(self, new_speedUCBMap):
    #     hour = 0
    #     for IDhour in new_speedUCBMap.keys():
    #         hour = int(IDhour.split(";")[1])
    #     #key_hour='hour_int'
    #     #if key_hour in new_speedUCBMap.keys():
    #     #    hour=new_speedUCBMap[key_hour]
    #     lengthUCB=self.roadLengthMap
    #     self.mabBus[hour].updateShadowBus(new_speedUCBMap, lengthUCB)
    #     #self.mabBus[hour].updateShadowBus(new_speedUCBMap)
    #     return hour


    # Initialize speed data for each link
    def initializeLinkEnergy1(self, fileName1):
        with open(fileName1, 'r') as f:
            f.readline() # skip the first line
            for line in f.readlines():
                result = line.split(",")
                roadID = int(result[0])
                roadType = int(float(result[2]))
                for i in range(int(self.args.SIMULATION_STOP_TIME * self.args.SIMULATION_STEP_SIZE//3600)+1):
                    backgroundSpeed = 35 if roadType==2 else 25
                    speedLength  = [backgroundSpeed]
                    self.initialLinkSpeedLength[i][roadID] = speedLength

    # Initialize link length data for each link
    def initializeLinkEnergy2(self, fileName2):   
        with open(fileName2, 'r') as f:
            f.readline()
            for line in f.readlines():
                result = line.split(",")
                roadID = int(result[0])
                roadLength = float(result[-1])
                roadType = int(float(result[2]))
                for i in range(int(self.args.SIMULATION_STOP_TIME * self.args.SIMULATION_STEP_SIZE//3600)+1):
                    backgroundSpeed =  35 if roadType==2 else 25
                    speedLength = [backgroundSpeed, roadLength]
                    self.initialLinkSpeedLength[i][roadID] = speedLength
                    self.roadLengthMap[roadID] = roadLength

        for i in range(int(self.args.SIMULATION_STOP_TIME * self.args.SIMULATION_STEP_SIZE//3600)+1):
            self.mab[i].warm_up(self.initialLinkSpeedLength[i])
            # self.mabBus[i].warm_up(self.initialLinkSpeedLength[i])

    def getRoadLengthMap(self):
        return self.roadLengthMap

